Book/php7/extensions_design/hooks.rst

```diff
@@ -0,0 +1,350 @@
+overwritten by an extension.
+
+::
+
+    static void (*original_zend_throw_exception_hook)(zval *ex);
+    void my_throw_exception_hook(zval *exception);
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        original_zend_throw_exception_hook = zend_throw_exception_hook;
+        zend_throw_exception_hook = my_throw_exception_hook;
+
+        return SUCCESS;
+    }
+
+If you implement this hook be aware that this hook is called regardless of
+whether the exception is caught or not. It can still be useful to temporarily
+store the exception here and then combine this with an implementation of the
+Error Handler hook to check if the exception was uncaught and caused the script
+to halt.
+
+Use-cases to implement this hook include debugging, logging and exception
+tracking.
+
+Hooking into eval()
+*******************
+
+PHPs ``eval`` is not an internal function but a special language construct. As
+such you cannot hook into it through ``zend_execute_internal`` or by
+overwriting its function pointer.
+
+Use cases for hooking into eval are not that many, you can use it for profiling
+or for security purposes. If you change its behavior be aware that other extensions
+may need eval. One example is Xdebug that uses it to execute breakpoint conditions.
+
+::
+
+    extern ZEND_API zend_op_array *(*zend_compile_string)(zval *source_string, char *filename);
+
+Hooking into the Garbage Collector
+**********************************
+
+PHPs Garbage Collector can be triggered explicitly when ``gc_collect_cycles()``
+is called or implicitly by the engine itself when the number of collectable
+objects reaches a certain theshold.
+
+To allow understanding of how the garabage collector works or to profile its
+performance, you can overwrite the function pointer hook that performs the
+garbage collection operation. Theoretically you can implement your own garbage
+collection algorithm here, but given other changes to the engine would probably
+be necessary this probably is not really feasible.
+
+::
+
+    int (*original_gc_collect_cycles)(void);
+
+    int my_gc_collect_cycles(void)
+    {
+        original_gc_collect_cycles();
+    }
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        original_gc_collect_cycles = gc_collect_cycles;
+        gc_collect_cycles = my_gc_collect_cycles;
+
+        return SUCCESS;
+    }
+
+Overwrite Interrupt Handler
+***************************
+
+The interrupt handler is called once when the executor global
+``EG(vm_interrupt)`` is set to 1. This is checked after every opcode.  The
+engine uses this hook to implement the PHP execution timeout via a signal
+handler that sets the interrupt to 1 after the timeout duration is reached.
+
+You can implement the interrupt helper yourself and trigger it by setting
+``EG(vm_interrupt) = 1`` in your extension.
+
+This can be helpful to implement debuggers, sampling profilers or your own
+timeout handling. By setting this hook you cannot accidently disable the
```

Comment:
For both debugging and profiling vm_interrupt is not great, because it will only interrupt at some point in the future. I would put the emphasis on signal handling here: You can defer processing of a signal that may occur at any time until a point where it is safe to interact with the engine (e.g. if you want to run PHP code).